### ISSUE TITLE
How to create datasource dynamically AFTER application start

### ISSUE THREAD

**Issue by ataft** [strongloop/loopback-datasource-juggler#1409]
Unlike issue #658, which shows how to create a datasource dynamically on application start in server.js, I'd like to create a datasource dynamically after the application is already started. The application would initially use the "memory" connector. The first time that the web application is accessed, it would prompt the administrator to enter the application database settings (i.e. where they want the loopback app database), then it would run autoupdate.js to create the application database. Then, the app would update datasources.json and use that going forward.

Basically, I don't want my clients to have to manually edit json config files after installing my app. I thought the following code would do it, before autoupdate on boot: 
```
  var loopback = require('loopback');

  var DataSource = require('loopback-datasource-juggler').DataSource;

  var sqlConnector = require('loopback-connector-mssql');

  var settings = {
    "host": "lenovo-w540",
    "port": 1433,
    "database": "appdb",
    "user": "username",
    "password": "password",
    "name": "appdb"
  };

  var dataSource = new DataSource(sqlConnector,settings);

  dataSource.connector = sqlConnector;           // Attach connector to dataSource
  sqlConnector.dataSource = dataSource;          // Hold a reference to dataSource

  app.dataSource('appdb', dataSource);

  var models = app.models();
  models.forEach(function(model) {
    model.attachTo(app.dataSources.appdb);
  });

```

The server starts up OK, but then throws this error after half a second or so:
```
Web server listening at: http://lenovo-w540:3000
Browse your REST API at http://lenovo-w540:3000/explorer
Connection fails: ConnectionError: Failed to connect to undefined:1433 - connect
 ECONNREFUSED 127.0.0.1:1433
It will be retried for the next request.

events.js:160
      throw er; // Unhandled 'error' event
      ^
ConnectionError: Failed to connect to undefined:1433 - connect ECONNREFUSED 127.
0.0.1:1433
    at Connection.<anonymous> (C:\apps\flexbiserver\node_modules\mssql\lib\tedio
us.js:378:25)
    at Connection.g (events.js:291:16)
    at emitOne (events.js:96:13)
    at Connection.emit (events.js:188:7)
    at Connection.socketError (C:\apps\flexbiserver\node_modules\tedious\lib\con
nection.js:531:14)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at emitErrorNT (net.js:1276:8)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
```

**Comment by ssh24** [comment]
@ataft This looks like a feature request for the [**mssql**](https://github.com/strongloop/loopback-connector-mssql) module. Could you please open a ticket on that repo and we will consider it for our backlog?

For now, only the [**mysql**](https://github.com/strongloop/loopback-connector-mssql)) module supports creating database dynamically. You need to specify a flag **createDatabase: true** on your `datasources.json` file and it would create the database if it does not exist.

Closing this issue.

**Comment by ataft** [comment]
@ssh24 No, this is not the case. My issue is similar to #658, which is also independent of the individual connectors/modules. Rather than using datasources.json, I want to present users of my app a web-based form that let's them enter the app datasource. After they do this, the app would run the autoupdate.js script to create the entities, then ideally move the datasource the user entered from in-memory to a JSON file.

Basically, I don't want my customers to have to edit JSON files in order to configure installations of my app. I have a completely automated on-prem installation that starts the app afterwards, but the main problem is that loopback requires a datasources.json file up front. I'm looking for a way around this.

**Comment by ssh24** [comment]
@ataft 
> the main problem is that loopback requires a datasources.json file up front

That is not true. You can start an app without having an empty `datasources.json` file. 

However, I think I know what the issue you are talking about. The app will throw an error on your boot-script files if they use the dynamic databases and will fail since the datasource is not defined. In that case you can move the script out of `./server/` directory initially and copy it once your user has added the database config dynamically.
